BUG: masked array proper deepcopies

tylerjereddy opened the pull request:
Fixes #22556
Fixes #21022

* add regression test and fix for gh-22556, where we were relying on the array `copy` arg to deepcopy a compound object type; I thought about performance issues here, but if you are already in the land of `object` and you are explicitly opting in to `deepcopy`, it seems like performance might be wishful thinking anyway

* add regression test and fix for gh-21022--this one was weirder but seems possible to sidestep by not trying to assign a shape of `()` to something that already has shape `()` and a non-writeable `shape` attribute

tylerjereddy committed:
BUG: masked array proper deepcopies

Fixes #22556
Fixes #21022

* add regression test and fix for gh-22556, where
we were relying on the array `copy` arg to deepcopy
a compound object type; I thought about performance issues
here, but if you are already in the land of `object` and
you are explicitly opting in to `deepcopy`, it seems like
performance might be wishful thinking anyway

* add regression test and fix for gh-21022--this one was
weirder but seems possible to sidestep by not trying
to assign a shape of `()` to something that already has
shape `()` and a non-writeable `shape` attribute

seberg commented on numpy/ma/core.py:
```diff
@@ -2870,7 +2870,13 @@ def __new__(cls, data=None, mask=nomask, dtype=None, copy=False,
+                        # the shapes were the same, so we can at least
+                        # avoid that path
+                        if data._mask.shape != data.shape:
+                            data._mask.shape = data.shape
```
juck :).  (To be clear, I guess this is fine, I generally hate the fact that this code sets the shape, but...)

seberg commented on numpy/ma/core.py:
```diff
@@ -6300,6 +6306,13 @@ def __deepcopy__(self, memo=None):
+        # copy semantics to do the right thing here
+        if self.dtype.type is np.object_:
+            for idx, _ in enumerate(copied):
+                copied[idx] = deepcopy(copied[idx])
```
Maybe just call deepcopy on `_data`?  You can guard it by `dtype.hasobject` instead.  If we put it here, that may copy it twice, but maybe that is better than the bug.

I don't trust that the code works right now for 2-D masked arrays.

seberg reviewed (commented):
Thanks for being brave and looking at MA issues ;).  I do think the deepcopy call needs to be on the `_data` attribute itself.  Maybe still not fully right then, but at least closer.

tylerjereddy committed:
MAINT: PR 23269 revisions

* guard masked array `__deepcopy__` special object
handling behind `dtype.hasobject`, based on reviewer
feedback

* masked array `__deepcopy__` for object type handling
now deepcopies from `self._data` directly, based on
reviewer feedback

* add a test case for 2D masked array object
deepcopies, since reviewer was not convinced this was
working

tylerjereddy commented:
I think I've pushed in the revisions from the first round of review. I added a test for a 2d masked array deepcopy as well.

seberg commented on numpy/ma/core.py:
```diff
@@ -6300,6 +6306,12 @@ def __deepcopy__(self, memo=None):
+        # contain compound types--you cannot depend on normal
+        # copy semantics to do the right thing here
+        if self.dtype.hasobject:
+            copied = deepcopy(self._data)
```
```suggestion
            copied._data = deepcopy(copied._data)
```
It looks like you are returning the data array only otherwise :).

tylerjereddy replied:
Looks like `_data` doesn't have a "setter," but maybe you can live with my revision.

Since the test suite was blind to it I also beefed up one of the tests to be sensitive to it and check the `_mask` deepcopy too.

seberg replied:
Oh, interesting, I thought it is a private attribute and fine to just replace it entirely.  This seems good, though.  Except that it should be `...` to support 0-D arrays.

(Might just do that and merge some time, feels a bit like a 95% fix, but this is MA, 95% fixes are probably about as good as it gets :))

tylerjereddy committed:
MAINT: PR 23269 revisions

* the deepcopy of masked array object type now also includes
the components beyond just `_data`; add a related test case

tylerjereddy committed:
BUG: PR 23269 revisions

* handle 0-D masked object array deepcopies, with
regression test, based on reviewer feedback

tylerjereddy commented:
I added a regression test for 0-D case and the suggested fix.

If there are any other concrete issues let me know. I suspect the "95 % fix" you're referring to is the fact that these are band-aids rather than "by design" perhaps.

seberg commented:
Lets give this is a go, thanks Tyler.  It certainly does a better job than before.

By 95%, I mean that this doesn't feel ideal and also I would be surprised if it was fully right, since masked arrays can e.g. even hold array subclasses as data and mask, and whether _those_ get deepcopied correctly here, I doubt.
